pkg/tbtc/tbtc.go

```diff
@@ -312,6 +312,109 @@ func Initialize(
 		}()
 	})
 
+	_ = chain.OnDepositSweepProposalSubmitted(func(event *DepositSweepProposalSubmittedEvent) {
+		go func() {
+			walletPublicKeyHash := event.Proposal.WalletPubKeyHash
+
+			if ok := deduplicator.notifyDepositSweepProposalSubmitted(
+				event.Proposal,
+			); !ok {
+				logger.Infof(
+					"deposit sweep proposal for wallet PKH [0x%x] "+
+						"has been already processed",
+					walletPublicKeyHash,
+				)
+				return
+			}
+
+			confirmationBlock := event.BlockNumber +
+				depositSweepProposalConfirmationBlocks
+
+			logger.Infof(
+				"observed deposit sweep proposal for wallet PKH [0x%x] "+
+					"at block [%v]; waiting for block [%v] to confirm",
+				walletPublicKeyHash,
+				event.BlockNumber,
+				confirmationBlock,
+			)
+
+			err := node.waitForBlockHeight(ctx, confirmationBlock)
+			if err != nil {
+				logger.Errorf(
+					"failed to confirm deposit sweep proposal for "+
+						"wallet PKH [0x%x]: [%v]",
+					walletPublicKeyHash,
+					err,
+				)
+				return
+			}
+
+			expiresAt, cause, err := chain.GetWalletLock(
+				walletPublicKeyHash,
+			)
+			if err != nil {
+				logger.Errorf(
+					"failed to get lock for wallet PKH [0x%x]: [%v]",
+					walletPublicKeyHash,
+					err,
+				)
+				return
+			}
+
+			// The event is confirmed if the wallet is locked due to a deposit
+			// sweep action.
+			if time.Now().Before(expiresAt) && cause == DepositSweep {
+				// Fetch all past deposit sweep proposal events starting from one
+				// confirmation period before the original event's block.
+				// If there was a chain reorg, the event we received could be
+				// moved to a block with a lower number than the one
+				// we received.
+				pastEvents, err := chain.PastDepositSweepProposalSubmittedEvents(
+					&DepositSweepProposalSubmittedEventFilter{
+						StartBlock: event.BlockNumber -
+							depositSweepProposalConfirmationBlocks,
+						WalletPublicKeyHash: walletPublicKeyHash,
+					},
+				)
+				if err != nil {
+					logger.Errorf(
+						"failed to get past deposit sweep proposal "+
+							"events for wallet PKH [0x%x]: [%v]",
+						walletPublicKeyHash,
+						err,
+					)
+					return
+				}
+
+				// Should not happen but just in case.
+				if len(pastEvents) == 0 {
+					logger.Errorf(
+						"no past deposit sweep proposal events for "+
+							"wallet PKH [0x%x]",
+						walletPublicKeyHash,
+					)
+					return
+				}
+
+				lastEvent := pastEvents[len(pastEvents)-1]
```

Comment:
I am not sure if this is correct in all cases:
- If there the same event was moved +- one block after the reorg, the deduplicator will handle it.
- If another event was emitted on a fork, the client will receive two events and they will pass the deduplicator. Two goroutines will be triggered. The goroutine from the earliest event should give up to let the "more recent" goroutine do its confirmation handling.

I believe the logic - as currently implemented - allows the "older" goroutine to proceed even though the most recent event on the chain is not sufficiently confirmed yet.

Reply:
We discussed it with @lukasz-zimnoch synchronously and came up with a conclusion that we are going to remove the last event check and leave the rest of the logic as-is. 

The reasoning behind `WalletCoordinator` is that it is just a coordination point based on the chain consensus. If enough clients received the event, they should follow it and execute a signature. All input parameters for that signature are validated, so even if there was a reorg and another event will land on the canonical chain later, the first signature will still be valid and approved by Bitcoin. The only reason we want to wait for 20 blocks after the coordination even is received is to make sure the right *type* of action is executed given different *types* of actions may have different lock times. We do not want to run into a situation when the majority of clients execute sweep with N blocks wallet lock time and the chain has M < N blocks wallet lock time because the canonical chain - as a result of a reorg - is supposed to execute e.g. redemption.

Reply:
Left that explanation in the code as well: a322ad60